content/blog/posts/o3de-motion-matching-2022-04.md

```diff
@@ -0,0 +1,179 @@
+
+##### 2. Narrow-phase
+
+Inside the narrow-phase, we iterate through the returned set of frames from the KD-tree, and evaluate and compare their cost against each other. The frame with the minimal cost is the best match that we transition to.
+
+Pseudo-code:
+
+```
+minCost = MAX;
+for_all (nearest frames found in the broad-phase)
+{
+	frameCost = 0.0
+	for_all (features)
+	{
+		frameCost += CalculateCost(feature);
+	}
+
+	if (frameCost < minCost)
+	{
+		// We found a better next matching frame
+		minCost = frameCost;
+		newBestMatchingFrame = currentFrame;
+	}
+}
+
+StartTransition(newBestMatchingFrame);
+```
+
+## Character trajectories
+
+Let's talk about trajectories. A trajectory is a useful way to encode where the character came from and its intention of where and how it wants to move in the future. We store a number of samples containing character positions as well as facing directions for a given time window. For example, 4 samples into the past across 0.7 seconds and 6 samples for the desired future trajectory across 1.2 seconds. The Euclidean distances between the samples tell us if the character is accelerating or coming to a stop. The sample locations indicate if it is moving forward or doing a left turn, and the facing direction gives us more information about if the character is strafing or moving backwards. Recording the past trajectory is easy as we know the current character transform, but how do we decide the future trajectory to generate, and why do we even need that?
+
+{{< image-width src="/images/blog/motion-matching/trajectory-history.png" width="100%" alt="Trajectory history samples showing the past positions and facing directions of the character." >}}
+
+## Controlling the character
+
+We talked about how the algorithm ensures that the animation is smooth but did not look into how we can actually guide and control our character yet - that is where the trajectory comes in. The motion matching algorithm not only takes a pose but also a desired future trajectory as input. It tries to synthesize an animation that matches the trajectory and makes the character walk onto, or close to, the desired path.
+
+{{< image-width src="/images/blog/motion-matching/script-canvas-graph.png" width="100%" alt="Script Canvas Graph for motion matching character control." >}}
+
+For an interactive game, we can use the gamepad's joystick position as a target to generate a desired future trajectory. By creating an exponential curve starting at the current character position going towards the character facing direction and then bend the curve towards the joystick's target position. This process is called trajectory prediction.
```

Comment:
Had trouble parsing this one. The second sentence seems like a fragment as written.

```suggestion
For an interactive game, we can use the gamepad's joystick position as a target to generate a desired future trajectory by creating an exponential curve starting at the current character position going towards the character facing direction, and then bend the curve towards the joystick's target position. This process is called trajectory prediction.
```